### PR TITLE
ADD: technical release to be able to send outputs with custom scripts (OP_RETURNs in particular)

### PR DESCRIPTION
**Pull request by Overtorment**
Preparing for a BIP-47 send feature, doing some refactorings to be able to send OP_RETURNs in transactions

@junderw can you pls take a look?

### CODE REVIEW

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 1]
```diff
@@ -1128,13 +1130,18 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
-      if (t.address.startsWith('bc1')) {
+      if (t.address && t.address.startsWith('bc1')) {
         // in case address is non-typical and takes more bytes than coinselect library anticipates by default
         t.script = { length: bitcoin.address.toOutputScript(t.address).length + 3 };
       }
+
+      if (t?.script?.hex) {
+        // setting length for coinselect lib manually as it is not aware of our field `hex`
+        t.script.length = t.script.hex.length / 2 - 4;
+      }
```
Comment (Overtorment) [inline class/wallets/abstract-hd-electrum-wallet.ts:1141]: @junderw my memory is a bit fuzzy here.  on line 1135 I have this `+3` compensating for something. I remember I calculated somehow this +3 but I dont remember how.

on line 1140 i added this `-4` empirically when running tests to nail the perfect  feerate per vbyte estimation.
can you provide some guidance please?

File: class/wallets/legacy-wallet.ts [hunk 2]
```diff
@@ -369,24 +369,21 @@ export class LegacyWallet extends AbstractWallet {
   }
 
   coinselect(
-    utxos: CoinSelectUtxo[],
-    targets: CoinSelectTarget[],
+    utxos: CreateTransactionUtxo[],
+    targets: CreateTransactionTarget[],
     feeRate: number,
-    changeAddress: string,
```
Comment (Overtorment) [inline class/wallets/legacy-wallet.ts:375]: funny but i dont remember why changeAddress was here. and it was never used

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/2xr4kuvfxitlgrenutolaguaj4?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1280)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/mwgy4oiidrwanttagtgb26faze?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1281)

File: class/wallets/abstract-hd-electrum-wallet.ts (hunk 1 above)
Comment (junderw) [reply class/wallets/abstract-hd-electrum-wallet.ts]: The only thing I can think of is trying to round up something that will eventually be divided by 4.

With integer math, adding 3 before dividing 4 is the same as dividing by 4 and rounding up in decimal math.

Comment (junderw) [reply class/wallets/abstract-hd-electrum-wallet.ts]: 4 weight units per vByte is the only 4 I can think of.

Comment (Overtorment) [reply class/wallets/abstract-hd-electrum-wallet.ts]: these 3 and 4 are unrelated. 3 is for other case.

i was trying to figure out what  script lenghts to use by looking at `node_modules/coinselect/utils.js`

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @marcosrdz] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/marcosrdz

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @junderw] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/junderw

**Review by limpbrains — APPROVED** [review]

**Review by marcosrdz — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @junderw] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/junderw

**Review by junderw — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!

**Pull request merged**